magento/magento-cloud-patches
# Magento Quality Patches - improvement of dev experience

### Description
- [MC-36033](https://jira.corp.magento.com/browse/MC-36033) Split ece-patches and magento-patches application
- [MC-35802](https://jira.corp.magento.com/browse/MC-35802) Dev experience - Implement a service to determine which patch is causing a conflict
- [MC-35824](https://jira.corp.magento.com/browse/MC-35802) Dev experience - Adapt tool for working with git-instance Magento

### Preconditions for test scenarios
- Check if your GitHub token has access to https://github.com/magento-mpi/magento-cloud-patches.git, https://github.com/magento/quality-patches.git, https://github.com/magento-mpi/ece-tools.git
- Add to composer.json 

```
 "repositories": {
        "repo": {
            "type": "composer",
            "url": "https://repo.magento.com"
        },
        "mcp": {
            "type": "vcs",
            "url": "https://github.com/magento-mpi/magento-cloud-patches.git"
        },
        "mqp": {
            "type": "vcs",
            "url": "https://github.com/magento/quality-patches.git"
        }
    },
    "require": {
        "magento/magento-cloud-patches": "dev-git-based as 1.0.99",
        "magento/quality-patches": "dev-MC-TEST as 1.0.999"
    },
```
- run 'composer update'
- commit changes

### Testing scenario 1 - Applying optional quality patch on a local and remote Cloud instance
- run './vendor/bin/ece-patches status'
- Expected: all patches have status 'Not applied'
- Add to .magento.env.yaml
```
 stage:
    build:
        QUALITY_PATCHES:
            - MCTEST-1001
```
- run on local instance './vendor/bin/ece-patches apply'
- run on local instance './vendor/bin/ece-patches status'
- Expected: all patches with Type 'Required' and Optional patch MCTEST-1001 have the status 'Applied'
- Push local changes to remote branch
- After redeploying login on a remote instance by SSH
- run on remote instance './vendor/bin/ece-patches status'
- Expected: all patches with Type 'Required' and Optional patch MCTEST-1001 have the status 'Applied'
- Expected: /var/log/patch.log file contains 
```
INFO: Patch MCTEST-1001 has been applied {"file":"/app/vendor/magento/quality-patches/src/../patches/MCTEST-1001__TEST_changes__2.3.5_ce.patch"}
INFO: Patch MCTEST-1001 has been applied {"file":"/app/vendor/magento/quality-patches/src/../patches/MCTEST-1001__TEST_changes__2.3.5_ee.patch"}
```
### Testing scenario 2 - test ece-patches 'apply' and 'revert' with QUALITY_PATCHES: '*' on local Cloud instance
- on local instance run './vendor/bin/ece-patches revert'
- on local instance run './vendor/bin/ece-patches status'
- Expected: all patches have status 'Not applied'
- Set to .magento.env.yaml
```
 stage:
    build:
        QUALITY_PATCHES:
            - '*'
```
- run on local instance './vendor/bin/ece-patches apply'
- run on local instance './vendor/bin/ece-patches status'
- Expected: all patches have the status 'Applied'
- on local instance run './vendor/bin/ece-patches revert'
- on local instance run './vendor/bin/ece-patches status'
- Expected: all patches have the status 'Not Applied'

### Testing scenario 3 - test magento-patches 'apply' and 'revert' on local Cloud instance
- on local instance run './vendor/bin/magento-patches apply MCTEST-1001'
- on local instance run './vendor/bin/magento-patches status'
- Expected: patch MCTEST-1001 has status 'Applied', other patches have status 'Not applied'
- run on local instance './vendor/bin/magento-patches revert MCTEST-1001 '
- run on local instance './vendor/bin/magento-patches status'
- Expected: patch MCTEST-1001 has status 'Not Applied', 
- on local instance run './vendor/bin/magento-patches apply MCLOUD-5650 MAGECLOUD-5069'
- on local instance run './vendor/bin/magento-patches status'
- Expected:  patch MCLOUD-5650 MAGECLOUD-5069 has status 'Applied', other patches have status 'Not applied'
- on local instance run './vendor/bin/magento-patches revert --all'
- on local instance run './vendor/bin/magento-patches status'
- Expected: all patches have the status 'Not Applied'
- on local instance run './vendor/bin/magento-patches apply' without arguments
- Expected:  error is displaying: 'Not enough arguments (missing: "list_of_patches").'
- on local instance run ./vendor/bin/magento-patches apply '*' and check result with './vendor/bin/magento-patches status'
- Expected: all patches have the status 'Applied'
- on local instance run './vendor/bin/magento-patches revert' without arguments
- Expected: nothing happened, there are no patches to revert
- on local instance run './vendor/bin/magento-patches status'
- Expected: all patches still have the status 'Applied'

### Testing scenario 4 - work on Magento git-instance
- Clone Magento 2.3.5 using git tag 'git clone --depth 1 --branch 2.3.5 https://github.com/magento/magento2ce.git'
- Add to composer.json 

```
 "repositories": {
        "repo": {
            "type": "composer",
            "url": "https://repo.magento.com"
        },
        "mcp": {
            "type": "vcs",
            "url": "https://github.com/magento-mpi/magento-cloud-patches.git"
        },
        "mqp": {
            "type": "vcs",
            "url": "https://github.com/magento/quality-patches.git"
        }
    },
    "require": {
        "magento/magento-cloud-patches": "dev-git-based as 1.0.99",
        "magento/quality-patches": "dev-MC-TEST as 1.0.999"
    },
```
- run 'composer update'
- run './vendor/bin/magento-patches status'
- Expected: see the table with available patches for Magento 2.3.5. All patches have status 'Not applied'
- run './vendor/bin/magento-patches apply MCLOUD-5837' and check status './vendor/bin/magento-patches status'
- Expected: MCLOUD-5837 has the status 'Applied'
- run './vendor/bin/magento-patches revert MCLOUD-5837' and check status './vendor/bin/magento-patches status'
- Expected: MCLOUD-5837 has the status 'Not applied'

### Release notes

For user-facing changes, add a meaningful release note. For example, see [Magento Cloud Patches release notes](https://devdocs.magento.com/cloud/release-notes/mcp-release-notes.html).

### Associated documentation updates

Add link to Magento DevDocs PR or Issue, if needed.



### Contribution checklist
